### PR TITLE
More proper implementation of batchnorm

### PR DESCRIPTION
**Pull request by ezyang**
```
    Handle batchnorm properly.
    
    Basic idea:
    - Pass buffers (marked as non-Variable tensors) as input variables to
      the trace.   Every buffer gets represented as an input variable
      to the trace, and we remember a correspondence of the underlying
      TH pointer and an input variable in the trace.
    - When we initially trace a function, we DO NOT record the buffers
      as edges.  This is so autograd doesn't have to know anything about buffers.
      If we ever turn buffers into requires_grad=False parameters, then
      this problem goes away.
    - When we primspec the buffer, NOW we reach into the cached buffers
      (now appropriately named) and gin up the buffer information we need.
    
    Other things:
    - CppOp execution is now supported (but lightly tested) using
      SimpleEval (thanks @apaszke!)
    
    Todo:
    - E2E tests need to have their hacks removed.
    - Figure out what is going on with backwards
    
    Signed-off-by: Edward Z. Yang <ezyang@fb.com>

----

    Lint after ToffeeIR, and subsequent fallout.
    
    I realized we weren't running the linter after ToToffeeIR, so
    I added a lint call.  It thus emerged that the current implementation
    was using "Unused" nodes that were not added to the graph,
    which was tripping the lint.  I fixed this a few ways:
    
    - BatchNorm and Conv primspecs were returning dead "unused" nodes
      for their (implicit) handle parameters.  I removed them because
      setOutputs handles this already, and a dead unused node which
      is not attached to the graph violates the "no dead nodes"
      invariant.
    
    - OK, but MaxPool actually needs to return a unused node for
      the output which supported by PyTorch but not Toffee; we need
      to error if subsequently in the trace this output is used.
      The new strategy is to have MaxPool's primspec return a None
      at the unused position, and then immediately *check* if there
      are any uses of that output.  If there are, that's an error!
    
    - I needed to adjust the Select invariant in the exporter loop:
      only if a Select node has *uses* is it mandatory for it to be
      defined in env.
    

```

### CODE REVIEW

**Comment by ezyang** [comment]
@apaszke, you'll want to specifically look at the "Handle batchnorm properly" commit. I also fixed a bug in your closure conversion code in "Don't DCE Params"

@zdevito, look at "Lint after ToffeeIR", that commit touches your rewritten export.cpp

File: torch/csrc/toffee/export.cpp [hunk 1]
```diff
@@ -27,47 +27,69 @@ std::string node_name(Node* n) {
...
     }
   };
   for (auto node : g->nodes()) {
     IR_IF(node, Select)
-      //selects are translated by multi-return nodes.
-      JIT_ASSERT(env.count(value) > 0);
+      // Selects are translated by multi-return nodes.
+      if (!node->uses().empty()) {
+        // Select nodes with NO uses are omitted from env
+        JIT_ASSERT(env.count(value) > 0);
```
Comment (zdevito) [inline torch/csrc/toffee/export.cpp:87]: I'd rather not start making conditions about when something is in the environment.  It tends to create constant action at a distance bugs when the assert that something is in the environment fails.

File: torch/csrc/toffee/export.cpp [hunk 2]
```diff
@@ -214,6 +238,11 @@ static void encodeGraph(toffee::GraphProto * p_g, std::shared_ptr<Graph> & g, co
       // of the select invariant
       continue;
     }
+    if (node->kind() == kConstant && !node->t(kValue).defined() && node->uses().empty()) {
+      // Undefined nodes never show up in ToffeeIR; they're just a tool
```
Comment (zdevito) [inline torch/csrc/toffee/export.cpp]: Please make these kUndefined -- In addition to always having a double check in the code, now all the comments are _calling_ them undefined nodes. This is a good sign it should be its own kind.

File: torch/csrc/toffee/export.cpp [hunk 3]
```diff
@@ -27,47 +27,69 @@ std::string node_name(Node* n) {
...
           "primspec produced too few outputs");
-        env[old] = ctx.graph->create(jit::kUnused);
+        if (!old->uses().empty()) {
+          throw std::runtime_error("In Toffee export, handles should be unused");
+        }
       } else {
-        if(!outputs[i]->hasType()) {
-          outputs[i]->setType(old->typeOption());
-          env[old] = outputs[i];
+        if (outputs[i]) {
```
Comment (zdevito) [inline torch/csrc/toffee/export.cpp:68]: I don't want to optionally putting mappings in the environment. For debugging, we need to distinguish between, "I never mapped this" and "I did map it, but to a value that should never be used", which was the intention of kUnused nodes. To maintain the invariant of having no dead nodes, can you just create a single kUnused node at the beginning of this pass and map everything whose primspec return None to that node. Then assert it is unused and delete it at the end of the pass?

File: torch/csrc/autograd/functions/toffee/batch_normalization.cpp [hunk 4]
```diff
@@ -25,23 +27,11 @@ jit::node_list BatchNormForward::primspec(PrimSpecContext* ctx, jit::node_list i
...
-  bn->addInput(sm);
-  bn->addInput(sv);
-
   if(this->training) {
-    g->appendNode(g->createSelect(bn, 1));
-    inplace_outputs.push_back(sm_start);
-    g->appendNode(g->createSelect(bn, 2));
-    inplace_outputs.push_back(sm_start+1);
+    g->appendNode(g->createSelect(bn, 1)->setType(bn->inputs().at(3)->type()));
+    inplace_outputs.push_back(3);
```
Comment (zdevito) [inline torch/csrc/autograd/functions/toffee/batch_normalization.cpp:27]: Is it safe to remove sm_start? We optionally add an input on line 16.

File: torch/csrc/toffee/export.cpp [hunk 5]
```diff
@@ -27,47 +27,69 @@ std::string node_name(Node* n) {
...
     for(auto & old : old_outputs) {
+      // TODO: what if there are multiple trailing handle outputs?  That is
+      // a serious invariant violation...
       if(i >= outputs.size()) {
         // primspecs do not deal with Handles at the moment
         // so we map handles to Unused nodes
         auto typ = old->typeOption();
         JIT_ASSERTM(typ && typ->kind() == jit::TypeKind::HandleType,
           "primspec produced too few outputs");
-        env[old] = ctx.graph->create(jit::kUnused);
```
Comment (zdevito) [inline torch/csrc/toffee/export.cpp:53]: Optionally mapping things in env is going to cause debugging issues down the line.

File: torch/csrc/autograd/functions/toffee/batch_normalization.cpp [hunk 6]
```diff
@@ -50,7 +40,7 @@ jit::node_list BatchNormForward::primspec(PrimSpecContext* ctx, jit::node_list i
   }
   bn->is_(jit::kInPlaceOutputs,std::move(inplace_outputs));
   ctx->batch_norm_count++;
-  return {orig_output,  g->create(jit::kUnused)};
```
Comment (zdevito) [inline torch/csrc/autograd/functions/toffee/batch_normalization.cpp:53]: I'd like to match the number of outputs to the number of outputs that the oringinal Cpp/PythonOp had so this should be {orig_output, nullptr}.

**Review by zdevito — COMMENTED** [review]
I don't have enough knowledge about the autograd engine to comment on the Buffer stuff, so I just added some comments on the Export pass. I'd like to always keep the environment complete rather than optionally adding nodes, and suggested some ways to accomplish that while maintaining IR invariants.

**Comment by zdevito** [comment]
As noted in the TODO, because this changes batchnorm output names, this should not be merged until the E2E tests are updated so that they run with the way batchnorm produces outputs.

**Comment by ezyang** [comment]
I am quite aware of the old env invariant (everything is always mapped). In fact, my first attempted refactor (not included here) attempted to maintain this invariant, but also added "explanations" for why a node was in the map but should not be used (because there are a number of different situations where an undefined node could be added to the map, and while I was debugging I needed to distinguish between them.)

But then I realized that all of this is unnecessary: instead of putting an unused node in the map (and thus kicking an error down the road to some later point in time), why not *eagerly* error at this point? If something goes wrong, you want to fail as *early* as possible. So instead of adding unused nodes to the map, just test if the corresponding old node had uses, and if it did, error. A good side effect of this is if a primspec claims that a PyTorch output is not used, but actually it *was* used in the net, at this point we can immediately error (and, though not implemented, we could easily give detailed information about what the offending operator was.) With the unused node strategy, you wouldn't notice this until some later lint pass, or by manual inspection of the graph.

File: torch/csrc/autograd/functions/batch_normalization.cpp [hunk 7]
```diff
@@ -100,8 +104,11 @@ auto BatchNormForward::apply(const variable_list& inputs) -> variable_list {
 
   auto outputs = as_tensor_list(std::move(output));
   return wrap_outputs(inputs, std::move(outputs), [&](FunctionFlags f) {
+    BatchNormParams params = *this;
+    params.cached_running_mean = running_mean;
+    params.cached_running_var = running_var;
     return std::make_shared<BatchNormBackward>(
-        f, *this, std::move(save_mean), std::move(save_std),
+        f, params, std::move(save_mean), std::move(save_std),
```
Comment (apaszke) [inline torch/csrc/autograd/functions/batch_normalization.cpp]: Use `std::move` here to avoid a copy

File: torch/csrc/autograd/functions/toffee/batch_normalization.cpp (hunk 4 above)
Comment (ezyang) [reply torch/csrc/autograd/functions/toffee/batch_normalization.cpp:27]: As per https://github.com/caffe2/caffe2/blob/master/caffe2/operators/spatial_batch_norm_op.cc bias is not optional. It looks like a comment of mine in convolution.cpp lead y'all astray, sorry about that.

File: torch/csrc/autograd/functions/toffee/batch_normalization.cpp (hunk 6 above)
Comment (ezyang) [reply torch/csrc/autograd/functions/toffee/batch_normalization.cpp:53]: If I understand correctly, this "extra output" on the original CppOp is the implicit handle output that is added to all operations which don't have transparently traceable backwards. Evidence: https://github.com/ezyang/pytorch/blob/jit/torch/csrc/autograd/functions/batch_normalization.cpp#L101 where we can see clearly that the outputs vector has only ONE output.

In my opinion, the presence of such a handle op is an implementation detail, and shouldn't need to be addressed in a primspec. Furthermore, the way C++ primspecs were defined was inconsistent with the way it was done in Python. For example: https://github.com/ezyang/pytorch/blob/jit/torch/autograd/_functions/tensor.py#L98 ; in fact, View at the IR level will also have a Handle output; here is a sample trace:

```
  %245.0 : Float(10, 5, 5), %245.1 : Handle = ^View((10, 5L, 5L))(%242.0), uses = [[%248.i0], []];
```

So, following the same reasoning as in the C++ primspec, you would have been obligated to also return a tuple `(orig_output, None)` in this Python op (and others.) This doesn't seem right.

File: torch/csrc/toffee/export.cpp (hunk 2 above)
Comment (ezyang) [reply torch/csrc/toffee/export.cpp]: OK I can do this

File: torch/csrc/jit/tracer.h [hunk 8]
```diff
@@ -128,16 +130,47 @@ inline Node* getValueTrace(const std::shared_ptr<TracingState>& state, const std
...
+  at::Tensor buffer;
+  TraceInput(std::shared_ptr<Variable> variable) : variable(variable) {}
+  TraceInput(at::Tensor buffer) : buffer(buffer) {}
+};
+
 // Start tracing, treating 'inputs' as inputs to the trace, which can be
 // varied on subsequent invocations of the trace.  Any other variables
 // will be treated as constants.
-inline std::shared_ptr<TracingState> enter(const variable_list& inputs) {
+inline std::shared_ptr<TracingState> enter(std::vector<TraceInput>&& trace_inputs) {
```
Comment (apaszke) [inline torch/csrc/jit/tracer.h]: There's no reason to take an rvalue reference in here - you don't modify/reuse `trace_inputs` for anything. const reference was good.

File: torch/csrc/autograd/functions/batch_normalization.cpp [hunk 9]
```diff
@@ -40,8 +40,12 @@ auto BatchNormForward::apply(const variable_list& inputs) -> variable_list {
   auto& input = inputs[0];
   auto& weight = inputs[1];
   auto& bias = inputs[2];
+  // TODO: undodgy-fy this
```
Comment (apaszke) [inline torch/csrc/autograd/functions/batch_normalization.cpp]: ? either elaborate what's wrong or remove the comment

File: torch/csrc/autograd/jit_closure.cpp [hunk 10]
```diff
@@ -284,6 +296,9 @@ std::unique_ptr<AutogradClosure> createAutogradClosure(Graph *graph) {
       } else {
         fn = std::make_shared<PythonCall>(value);
       }
+    // TODO: not sure if this works...
+    IR_ELSEIFM(CppOp)
+      fn = std::make_shared<SimpleEval>(value->fn);
```
Comment (apaszke) [inline torch/csrc/autograd/jit_closure.cpp]: Can you change that to `fn = value->fn`?

File: torch/csrc/jit/python_tracer.cpp [hunk 11]
```diff
@@ -120,17 +121,22 @@ PyObject * THPTracer_enter(PyObject *_unused, PyObject *args)
...
+      trace_inputs.emplace_back(((THPVariable*)input_obj)->cdata);
+    } else if (THPModule_isTensor(input_obj)) {
+      trace_inputs.emplace_back(torch::createTensor(input_obj));
+    } else {
+      throw std::runtime_error("non-variable, non-tensor input");
+    }
   }
 
-  return THPTracingState_Wrap(tracer::enter(inputs));
+  return THPTracingState_Wrap(tracer::enter(std::move(trace_inputs)));
```
Comment (apaszke) [inline torch/csrc/jit/python_tracer.cpp:139]: no need to move this once you change it to const reference

File: torch/csrc/jit/tracer.h [hunk 12]
```diff
@@ -43,6 +43,8 @@ struct TracingState : public std::enable_shared_from_this<TracingState> {
     , active(false) {}
 
   std::shared_ptr<Graph> graph;
+  // void* is an unsafe TH.  NON-OWNING, so it might get invalidated
+  std::unordered_map<void*, Node*> buffer_map;
```
Comment (apaszke) [inline torch/csrc/jit/tracer.h]: We might want to actually change these to owning pointers. The whole point of buffers is that they're persistent, so it shouldn't increase mem usage

File: torch/csrc/toffee/export.cpp [hunk 13]
```diff
@@ -134,7 +137,7 @@ std::shared_ptr<Graph> ToToffeeIR(std::shared_ptr<Graph>& g) {
   for (auto output : g->outputs()) {
     ctx.graph->registerOutput(env.at(output));
   }
-  return std::move(ctx.graph);
+  return std::move(out_graph);
```
Comment (apaszke) [inline torch/csrc/toffee/export.cpp]: Don't move it. It prevents copy-elision

File: torch/csrc/autograd/functions/toffee/batch_normalization.cpp (hunk 6 above)
Comment (zdevito) [reply torch/csrc/autograd/functions/toffee/batch_normalization.cpp:53]: We need to talk about Handles. I believe we are out of sync about how we are going to handle them. To me, it is not an implementation detail, but something that will eventually be handled by primspecs when recognizing full backward passes.

**Comment by zdevito** [comment]
@ezyang I like the change to have something error if someone tries to use a node that should not be used.  But I _still_ want the error to distinguish from the bug of not mapping stuff. I'd like the error to say "unused not is actually used" or "node never mapped".  These are all assert, but important asserts as the code gets more compllicated.
